### PR TITLE
[incremental] Add support for eval always queries

### PR DESCRIPTION
**Pull request by wesleywiser**
Part of #45238

### CODE REVIEW

File: src/librustc/dep_graph/graph.rs [hunk 1]
```diff
@@ -782,6 +812,24 @@ impl CurrentDepGraph {
         }
     }
 
+    pub(super) fn push_untracked_task(&mut self, key: DepNode) {
```
Comment (michaelwoerister) [inline src/librustc/dep_graph/graph.rs]: You should be able to make the `push` and the `pop` methods here private.

File: src/librustc/dep_graph/graph.rs [hunk 2]
```diff
@@ -812,7 +860,7 @@ impl CurrentDepGraph {
                     reads.push(source);
                 }
             }
-            Some(&mut OpenTask::Ignore) | None => {
+            Some(&mut OpenTask::Ignore) | Some(&mut OpenTask::Untracked { .. }) | None => {
```
Comment (michaelwoerister) [inline src/librustc/dep_graph/graph.rs]: Since this line is getting a little long, it would be more readable to have each arm on its own line.

**Review by michaelwoerister — COMMENTED** [review]
Looking good, except for the result fingerprinting being missing. Since we essentially disable dependency tracking for this tasks its even more important for them to be fingerprinted. 

If you add the fingerprint, however, I guess there'll be a lot of code duplication between `with_untracked_task` and `with_task`. Maybe one can be implemented in the terms of the other.

I'm also a bit skeptical about the term `untracked` since it rather suggests the functionality that we already have with `ignore`.

**Comment by michaelwoerister** [comment]
Hm, it looks like implementing `with_untracked_task` in terms `with_task` gets rather contrived. I had hoped that we would be able to do something like:

```rust

pub fn with_xxx_task<C, A, R, HCX>(&self,
                                   key: DepNode,
                                   cx: C,
                                   arg: A,
                                   task: fn(C, A) -> R)
                                   -> (R, DepNodeIndex) 
{
    self.with_task(key, cx, arg, |cx, arg| {
        // Explicitly add the read to the Krate node
        self.read(DepNode::Krate);

        // Execute the task without recording any other reads
        self.with_ignore(|| {
            task(cx, arg)
        })     
    })
}
```
But `with_task` takes a function, not a closure, so we cannot capture the `task` parameter. If you can come up with a way to make this work nonetheless, that'd be nice. Then you wouldn't even need to modify `CurrentDepGraph`. Otherwise I suggest to make a generalized, private version of `with_task` -- which allows to specify somehow which `push` and `pop` methods it will invoke -- and call that with different parameters from `with_task` and `with_untracked_task`.
Otherwise I sugg

**Comment by wesleywiser** [comment]
Thanks for the review! What about something like "crate-wide query"? I'm also fine just changing it to "eval-always".

**Comment by michaelwoerister** [comment]
Yeah, let's stick to "eval-always". It's not a pretty name but it conveys the underlying concept well enough.

**Comment by wesleywiser** [comment]
@michaelwoerister I fixed up the commit per your feedback.

**Comment by michaelwoerister** [comment]
@wesleywiser Yes, that looks very good now! If you want to continue working on this, I suggest just appending to this PR.

**Comment by wesleywiser** [comment]
@michaelwoerister Thanks! Pushed three new commits.

**Comment by michaelwoerister** [comment]
Looks very good!

I'm seeing this error on travis:
```
thread 'rustc' panicked at 'assertion failed: !data.colors.borrow().contains_key(&key)', /checkout/src/librustc/dep_graph/graph.rs:220:12
```

This suggests that we are re-running queries although they've already been marked as green. This is because we are not doing the whole `try_mark_green_and_read()` spiel for eval-always. 

Thinking about it some more, there should really be no reason to treat an eval-always query different from a regular one, except for the changes you already implemented in `graph.rs`. So instead of modifying `try_get()`, could you just do `if dep_node.is_eval_always() {...}` in `force()`?:
https://github.com/rust-lang/rust/blob/b2478052f88db8c8526ee2dc4a382da91eefc76c/src/librustc/ty/maps/plumbing.rs#L433-L442

That should also not introduce any code duplication.

Let me know if you have any questions about this change of strategy.

**Comment by wesleywiser** [comment]
Hmmm... I can't seem to repro that. `./x.py test` succeeds on my machine. 

Regardless, I can certainly revert the changes to `try_get_with()` but I'm not entirely clear what you're proposing in regards to `force()`. Can you elaborate on that?

**Comment by michaelwoerister** [comment]
> Hmmm... I can't seem to repro that. ./x.py test succeeds on my machine.

You are probably building with debug assertions disabled. I generally recommend setting `debug-assertions = true` in your `config.toml`. (and `debuginfo-lines = true` for better backtraces).

I imagine the updated `force()` method to look like this:
```rust
profq_msg!(tcx, ProfileQueriesMsg::ProviderBegin); 
let res = tcx.cycle_check(span, Query::$name(key), || { 
    tcx.sess.diagnostic().track_diagnostics(|| { 
        if dep_node.is_eval_always() {
            tcx.dep_graph.with_eval_always_task(dep_node, 
                                                tcx, 
                                                key, 
                                                Self::compute_result) 
        } else {
            tcx.dep_graph.with_task(dep_node, 
                                    tcx, 
                                    key, 
                                    Self::compute_result) 
        }
    }) 
})?; 
profq_msg!(tcx, ProfileQueriesMsg::ProviderEnd);
```
This way eval-always tasks are treated the same as regular with the only exception of what reads are recorded from them.

**Comment by wesleywiser** [comment]
Oh got it. That makes sense.

> I generally recommend setting debug-assertions = true in your config.toml. (and debuginfo-lines = true for better backtraces).

Thanks! I'll do that.

**Comment by wesleywiser** [comment]
@michaelwoerister Done

**Comment by michaelwoerister** [comment]
Awesome, thank you so much @wesleywiser!

@bors r+

**Comment by wesleywiser** [comment]
Thanks for all the help @michaelwoerister!!

**Comment by michaelwoerister** [comment]
You're welcome!

**Pull request merged**